keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,867 @@
+      // reserved
+      ServiceExportInitialized ServiceExportConditionType = "Initialized"
+      // ServiceExportExported means that the service referenced by this
+      // service export has been synced to all clusters in the supercluster
+      ServiceExportExported ServiceExportConditionType = "Exported"
+}
+
+// ServiceExportCondition contains details for the current condition of this
+// service export.
+//
+// Once [#1624](https://github.com/kubernetes/enhancements/pull/1624) is
+// merged, this will be replaced by metav1.Condition.
+type ServiceExportCondition struct {
+        Type ServiceExportConditionType `json:"type"`
+        // Status is one of {"True", "False", "Unknown"}
+        Status corev1.ConditionStatus `json:"status"`
+        // +optional
+        LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+        // +optional
+        Reason *string `json:"reason,omitempty"`
+        // +optional
+        Message *string `json:"message,omitempty"`
+}
+```
+```yaml
+apiVersion: multicluster.k8s.io/v1alpha1
+kind: ServiceExport
+metadata:
+  name: my-svc
+  namespace: my-ns
+status:
+  conditions:
+  - type: Initialized
+    status: "True"
+    lastTransitionTime: "2020-03-30T01:33:51Z"
+  - type: Exported
+    status: "True"
+    lastTransitionTime: "2020-03-30T01:33:55Z"
+```
+
+`ServiceExports` will be created within the cluster and namespace that the
+service resides in and are name-mapped to the service for export - that is, they
+reference the `Service` with the same name as the export. If multiple clusters
+within the supercluster have `ServiceExports` with the same name and namespace,
+these will be considered the same service and will be combined at the
+supercluster level.
+
+This requires that within a supercluster, a given namespace is governed by a
+single authority across all clusters. It is that authority’s responsibility to
+ensure that a name is shared by multiple services within the namespace if and
+only if they are instances of the same service.
+
+Most information about the service, including ports, backends and topology, will
+continue to be stored in the Service object, which is name mapped to the service
+export.
+
+### Exported Service Behavior Expectations
+
+#### SuperclusterIP
+
+When a `ServiceExport` is created, an IP address is reserved and assigned to
+this supercluster `Service`. This IP may be supercluster-wide, or assigned on a
+per-cluster basis. Requests to the corresponding IP from within a given cluster
+will route to endpoint addresses for the aggregated Service. 
+
+Note: this doc does not discuss `NetworkPolicy`, which cannot currently be used
+to describe a policy that applies to a multi-cluster service.
+
+#### DNS
+
+When a `ServiceExport` is created, this will cause a domain name for the
+multi-cluster service to become accessible from within the supercluster. The
+domain name will be
+`<service-export-name>.<service-export-namespace>.svc.supercluster.local`.
```

Comment:
Need to determine:
1. If the supercluster.local suffix can be overridden, and if so on a per-cluster or supercluster basis.
2. What happens if the user happens to have used "supercluster.local" as their suffix for their cluster.

For #2, maybe it should not be .svc. and instead something else? Then there would be no potential for conflict. .ssvc.cluster.local? .super.cluster.local?

Reply:
Yeah this is a tough one. I think this needs further discussion. In an ideal world we would be able to reserve a name here, but that is not our world :). First thoughts:

1. I think ideally we'd have some consistent standard name here, and you could create an alias if needed. I don't see why this couldn't be per-cluster as with the VIP - you'd likely want it to be supercluster level, but that's up to the implementation since each cluster consumes a service independently based solely on local information.

2. Maybe something.cluster.local is a solution. It would avoid conflicts, but semantically it feels like a big stretch for cluster.local as it's explicitly not cluster local.

Reply:
https://github.com/kubernetes/dns/blob/master/docs/specification.md for reference

Reply:
Agree with all points. I'll absolutely add DNS spec as a graduation criteria.

In the meantime, it seems like a few questions have come up repeatedly in discussions around this KEP:

- Should the name be fixed or configurable? fixed makes life easier, but users may want customization. 
- Should supercluster services use a new suffix or live under cluster.local (e.g. super.cluster.local)?
- What should the structure of a supercluster address be?

I don't think we need the svc literal in there, but I don't think we can say at this point that there will never be supercluster names for non-services (though I don't know what they would be). The main reason this proposal has kept the literal is for familiarity with cluster.local names, which seems like a good thing to maintain. I'm all for restructuring in-cluster DNS and supercluster DNS should follow suit, but it seems risky to develop supercluster DNS improvements independently

Reply:
The main failing of `svc.<suffix>` is that it is used in search paths (which I do not expect here), which makes for 2 user-controlled tokens which can be ambiguous.  It's also WRONG from a hierarchy point of view (services are smaller than namespaces), but I think I'd be OK to keep it for consistency.

Reply:
Completely agree. Consistency is the main driver, but fixing the hierarchy in general would be great and MCS should adopt any improvements